src/rpc/RpcHandler.ts

```diff
@@ -1,24 +1,58 @@
 import P2PManager from "../P2PManager";
 import ATransport from "../transport/ATransport";
-import Rpc, { serializeRpc } from "./Rpc";
+import Rpc, { createMessageContent, serializeRpc } from "./Rpc";
+
+type UnsignedRpc = Omit<Rpc, "signature">;
 
 class RpcHandler {
-    rpc: Rpc;
-    p2pManager: P2PManager;
-    constructor(rpc: Rpc, p2pManager: P2PManager) {
-        this.rpc = rpc;
+    private readonly rpcPayload: UnsignedRpc;
+    private readonly p2pManager: P2PManager;
+    private signedRpcPromise: Promise<Rpc> | null = null;
+
+    constructor(rpcPayload: UnsignedRpc, p2pManager: P2PManager) {
+        this.rpcPayload = rpcPayload;
         this.p2pManager = p2pManager;
     }
 
-    public broadcast() {
-        this.p2pManager.broadcastRpc(serializeRpc(this.rpc));
+    private async buildSignedRpc(): Promise<Rpc> {
+        if (!this.signedRpcPromise) {
+            this.signedRpcPromise = (async () => {
```

Comment:
Imo better to `await` here and have the `signedRpc` instead of `signedRpcPromise`